test/boost/replicator_test.cc

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) 2025-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#include "test/lib/mutation_source_test.hh"
+#include "test/lib/random_utils.hh"
+#include "test/lib/cql_test_env.hh"
+#include "test/lib/test_utils.hh"
+#include "test/lib/log.hh"
+
+#include "utils/replicator.hh"
+#include "utils/throttle.hh"
+
+#undef SEASTAR_TESTING_MAIN
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/util/defer.hh>
+#include <fmt/format.h>
+#include <fmt/std.h>
+#include <map>
+#include <optional>
+
+
+BOOST_AUTO_TEST_SUITE(replicator_test)
+
+using map_type = std::map<int, std::optional<int>>;
+using key_setter = std::pair<int, int>;
+
+static void require_multishard() {
+    tests::require_greater(smp::count, 1);
+}
+
+// last-write-wins semantics for mutations.
+struct map_container : public seastar::peering_sharded_service<map_container> {
+    map_type _m;
+    std::function<void()> _on_apply = [] {};
+    uint64_t _failures = 0;
+
+    void apply_locally(const map_type& m) {
+        _on_apply();
+        apply(_m, m);
+    }
+
+    void apply_locally(std::pair<int, int> kv) {
+        map_type src;
+        apply(src, kv);
+        apply_locally(src);
+    }
+
+    static void apply(map_type& dst, const map_type& src) {
+        for (auto&& e : src) {
+            if (e.second) {
+                dst[e.first] = *e.second;
+            }
+        }
+    }
+
+    static void apply(map_type& dst, const key_setter& src) {
+        apply(dst, map_type{std::make_pair(src.first, std::make_optional(src.second))});
```

Comment:
I guess that would be optimized, but to me it makes more sense to reverse the calling structure and have apply(map_type, const map_type&) call this function for every pair in `src` that have its value engaged.